nx/lib/nx.ex

```diff
@@ -7313,6 +7353,89 @@ defmodule Nx do
     slice(tensor, start_indices, lengths, opts)
   end
 
+  @doc """
+  Puts the given slice into the given tensor at the given
+  start indices.
+
+  The given slice shape must be less than or equal to the
+  shape of the given tensor. All start indices must be less
+  than their respective dimensions.
+
+  ## Examples
+
+      iex> Nx.put_slice(Nx.tensor([0, 1, 2, 3, 4]), Nx.tensor([5, 6]), [2])
+      #Nx.Tensor<
+        s64[5]
+        [0, 1, 5, 6, 4]
+      >
+
+      iex> Nx.put_slice(Nx.tensor([[1, 2, 3], [4, 5, 6]]), Nx.tensor([[7, 8], [9, 10]]), [1, 2])
+      #Nx.Tensor<
+        s64[2][3]
+        [
+          [1, 7, 8],
+          [4, 9, 10]
+        ]
+      >
+
+      iex> Nx.put_slice(Nx.tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]), Nx.tensor([[7.0, 8.0, 9.0], [10.0, 11.0, 12.0]]), [2, 2])
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [7.0, 8.0, 9.0],
+          [10.0, 11.0, 12.0]
+        ]
+      >
+
+      iex> Nx.put_slice(Nx.tensor([[1, 2, 3], [4, 5, 6]]), Nx.tensor([[10.0, 11.0]]), [Nx.tensor(0), Nx.tensor(2)])
+      #Nx.Tensor<
+        f32[2][3]
+        [
+          [1.0, 10.0, 11.0],
+          [4.0, 5.0, 6.0]
+        ]
+      >
+  """
+  def put_slice(tensor, slice, start_indices) when is_list(start_indices) do
+    %T{shape: shape, names: names, type: type} = tensor = to_tensor(tensor)
+    %T{shape: slice_shape, names: slice_names, type: slice_type} = slice = to_tensor(slice)
+
+    output_type = binary_type(type, slice_type)
+
+    start_indices =
+      start_indices
```

Comment:
We can probably extract this logic to some place we can reuse. Also please add a TODO for us to use `Enum.with_index/3` in Elixir v1.12. :)